### PR TITLE
Gerudo1 OK and mostly documented

### PR DESCRIPTION
**Pull request by EllipticEllipsis**
White-clad Gerudo, including gate guards, gate operator, GTG guard, Horseback Archery. Turns out they have 3 different hairstyles.

All functions decompiled, functions named (although some may be inaccurate due to incomplete information), most variables named. Defines made for state flags, enums made for types, hairstyles, limbs. Asschest briefly documented.

Large, but not too large actor, with some odd functions and inconsistent programming styles.

### CODE REVIEW

File: src/overlays/actors/ovl_En_Ge1/z_en_ge1.c [hunk 1]
```diff
@@ -26,95 +73,744 @@ const ActorInit En_Ge1_InitVars = {
...
+    this->animFunc = EnGe1_CueUpAnimation;
+    this->actor.unk_1F = 6;
+    Actor_SetScale(&this->actor, 0.01f);
+
+    // In Gerudo Valley
+    if (globalCtx->sceneNum == SCENE_SPOT09) {
+        this->actor.uncullZoneForward = 1000.0f;
+    } else {
+        this->actor.uncullZoneForward = 1200.0f;
+    }
```
Comment (petrie911) [inline src/overlays/actors/ovl_En_Ge1/z_en_ge1.c:122]:
```suggestion
    this->actor.uncullZoneForward = (globalCtx->sceneNum == SCENE_SPOT09) ? 1000.0f : 1200.0f;
```
Ternary possible here?

File: src/overlays/actors/ovl_En_Ge1/z_en_ge1.c [hunk 2]
```diff
@@ -26,95 +73,744 @@ const ActorInit En_Ge1_InitVars = {
...
+}
+
+void EnGe1_WatchForAndSensePlayer(EnGe1* this, GlobalContext* globalCtx) {
+    s16 angleDiff = this->actor.yawTowardsLink - this->actor.shape.rot.y;
+
+    if ((this->actor.xzDistToLink < 50.0f) || ((ABS(angleDiff) <= 0x4300) && (this->actor.xzDistToLink < 400.0f))) {
+        EnGe1_SpotPlayer(this, globalCtx);
+    }
+
+    if (this->collider.base.acFlags & 2) {
```
Comment (petrie911) [inline src/overlays/actors/ovl_En_Ge1/z_en_ge1.c:312]:
```suggestion
    if (this->collider.base.acFlags & AC_HIT) {
```

File: src/overlays/actors/ovl_En_Ge1/z_en_ge1.c [hunk 3]
```diff
@@ -26,95 +73,744 @@ const ActorInit En_Ge1_InitVars = {
...
+        EnGe1_SetTalkAction(this, globalCtx, 0x6013, 100.0f, EnGe1_RefuseOpenNoCard_GTGGuard);
+    }
+}
+
+// Gate Operator functions
+
+void EnGe1_WaitGateOpen_GateOp(EnGe1* this, GlobalContext* globalCtx) {
+    this->stateFlags |= GE1_STATE_TALKING;
+
+    if ((func_8010BDBC(&globalCtx->msgCtx) == 5) && (func_80106BC8(globalCtx) != 0)) {
```
Comment (petrie911) [inline src/overlays/actors/ovl_En_Ge1/z_en_ge1.c:416]:
```suggestion
    if ((func_8010BDBC(&globalCtx->msgCtx) == 5) && func_80106BC8(globalCtx)) {
```
Pretty sure this is boolean.

File: src/overlays/actors/ovl_En_Ge1/z_en_ge1.c [hunk 4]
```diff
@@ -26,95 +73,744 @@ const ActorInit En_Ge1_InitVars = {
...
+
+    if (func_8002F334(&this->actor, globalCtx)) {
+        this->actor.flags = this->actor.flags & ~0x10000;
+        this->actionFunc = EnGe1_WaitTillItemGiven_Archery;
+    }
+    if (this->stateFlags & GE1_STATE_GIVE_QUIVER) {
+        switch (CUR_UPG_VALUE(UPG_QUIVER)) {
+            //! @bug Asschest: the compiler inserts a default assigning *(sp+0x24) to getItemId, which is junk data left
+            //! over from the previous function run in EnGe1_Update, namely EnGe1_CueUpAnimation. The top stack variable
+            //! in that function is &this->skelAnima = thisx + 198, and depending on where this loads in memory, the
```
Comment (petrie911) [inline src/overlays/actors/ovl_En_Ge1/z_en_ge1.c:542]:
```suggestion
            //! in that function is &this->skelAnime = thisx + 198, and depending on where this loads in memory, the
```

File: src/overlays/actors/ovl_En_Ge1/z_en_ge1.c [hunk 5]
```diff
@@ -26,95 +73,744 @@ const ActorInit En_Ge1_InitVars = {
...
+    if (func_8002F194(&this->actor, globalCtx)) {
+        this->actionFunc = EnGe1_BeginGiveItem_Archery;
+        this->actor.flags &= ~0x10000;
+    } else {
+        func_8002F2CC(&this->actor, globalCtx, 200.0f);
+    }
+}
+
+void EnGe1_TalkTooPoor_Archery(EnGe1* this, GlobalContext* globalCtx) {
+    if ((func_8010BDBC(&globalCtx->msgCtx) == 5) && (func_80106BC8(globalCtx) != 0)) {
```
Comment (petrie911) [inline src/overlays/actors/ovl_En_Ge1/z_en_ge1.c:567]:
```suggestion
    if ((func_8010BDBC(&globalCtx->msgCtx) == 5) && func_80106BC8(globalCtx)) {
```

File: src/overlays/actors/ovl_En_Ge1/z_en_ge1.c [hunk 6]
```diff
@@ -26,95 +73,744 @@ const ActorInit En_Ge1_InitVars = {
...
+        func_80106CCC(globalCtx);
+    } else if ((this->skelAnime.curFrame == 15.0f) || (this->skelAnime.curFrame == 19.0f)) {
+        Audio_PlayActorSound2(&this->actor, NA_SE_IT_HAND_CLAP);
+    }
+}
+
+void EnGe1_SetupOpenGate_GateOp(EnGe1* this, GlobalContext* globalCtx) {
+    this->stateFlags |= GE1_STATE_TALKING;
+
+    if ((func_8010BDBC(&globalCtx->msgCtx) == 5) && (func_80106BC8(globalCtx) != 0)) {
```
Comment (petrie911) [inline src/overlays/actors/ovl_En_Ge1/z_en_ge1.c:447]:
```suggestion
    if ((func_8010BDBC(&globalCtx->msgCtx) == 5) && func_80106BC8(globalCtx)) {
```

File: src/overlays/actors/ovl_En_Ge1/z_en_ge1.c [hunk 7]
```diff
@@ -26,95 +73,744 @@ const ActorInit En_Ge1_InitVars = {
...
+void EnGe1_RefuseEntryTooPoor_GTGGuard(EnGe1* this, GlobalContext* globalCtx) {
+    if (func_8010BDBC(&globalCtx->msgCtx) == 2) {
+        this->actionFunc = EnGe1_CheckForCard_GTGGuard;
+        EnGe1_SetAnimationIdle(this);
+    }
+}
+
+void EnGe1_OfferOpen_GTGGuard(EnGe1* this, GlobalContext* globalCtx) {
+    this->stateFlags |= GE1_STATE_TALKING;
+    if ((func_8010BDBC(&globalCtx->msgCtx) == 4) && (func_80106BC8(globalCtx) != 0)) {
```
Comment (petrie911) [inline src/overlays/actors/ovl_En_Ge1/z_en_ge1.c:372]:
```suggestion
    if ((func_8010BDBC(&globalCtx->msgCtx) == 4) && func_80106BC8(globalCtx)) {
```

File: src/overlays/actors/ovl_En_Ge1/z_en_ge1.c [hunk 8]
```diff
@@ -26,95 +73,744 @@ const ActorInit En_Ge1_InitVars = {
...
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Ge1/func_80A32190.s")
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Ge1/EnGe1_Update.s")
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Ge1/func_80A323B0.s")
+static ColliderCylinderInit sCylinderInit = {
+    { COLTYPE_UNK10, 0x00, 0x11, 0x39, 0x10, COLSHAPE_CYLINDER },
+    { 0x00, { 0x00000000, 0x00, 0x00 }, { 0x00000702, 0x00, 0x00 }, 0x00, 0x01, 0x01 },
+    { 20, 40, 0, { 0, 0, 0 } },
+};
```
Comment (petrie911) [inline src/overlays/actors/ovl_En_Ge1/z_en_ge1.c:81]: Needs updating to the new format.

**Review by fig02 — APPROVED** [review]
changes made

File: src/overlays/actors/ovl_En_Ge1/z_en_ge1.c [hunk 9]
```diff
@@ -46,95 +65,732 @@ static ColliderCylinderInit D_80A32770 = {
...
+            if (INV_CONTENT(SLOT_BOW) == ITEM_NONE) {
+                Actor_Kill(&this->actor);
+                return;
+            }
+            this->actor.targetMode = 3;
+            this->hairstyle = GE1_HAIR_BOB;
+            // Horsback archery Gerudo EVENT_INF(0) =
+            osSyncPrintf(VT_FGCOL(CYAN) "やぶさめ ゲルド EVENT_INF(0) = %x\n" VT_RST, gSaveContext.eventInf[0]);
+
+            if ((gSaveContext.eventInf[0] & 0x100)) {
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_Ge1/z_en_ge1.c:153]:
```suggestion
            if (gSaveContext.eventInf[0] & 0x100) {
```

File: src/overlays/actors/ovl_En_Ge1/z_en_ge1.c [hunk 10]
```diff
@@ -46,95 +65,732 @@ static ColliderCylinderInit D_80A32770 = {
...
+    EnGe1* this = THIS;
+
+    Collider_DestroyCylinder(globalCtx, &this->collider);
+}
+
+s32 EnGe1_SetTalkAction(EnGe1* this, GlobalContext* globalCtx, u16 textId, f32 arg3, EnGe1ActionFunc actionFunc) {
+    if (func_8002F194(&this->actor, globalCtx)) {
+        this->actionFunc = actionFunc;
+        this->animFunc = EnGe1_StopFidget;
+        this->stateFlags &= ~(GE1_STATE_IDLE_ANIM);
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_Ge1/z_en_ge1.c:185]:
```suggestion
        this->stateFlags &= ~GE1_STATE_IDLE_ANIM;
```

File: src/overlays/actors/ovl_En_Ge1/z_en_ge1.c [hunk 11]
```diff
@@ -46,95 +65,732 @@ static ColliderCylinderInit D_80A32770 = {
...
+}
+
+void EnGe1_SetupOpen_GTGGuard(EnGe1* this, GlobalContext* globalCtx) {
+    if ((func_8010BDBC(&globalCtx->msgCtx) == 5) && func_80106BC8(globalCtx)) {
+        this->actionFunc = EnGe1_Open_GTGGuard;
+        Animation_Change(&this->skelAnime, &D_0600A048, 1.0f, 0.0f, Animation_GetLastFrame(&D_0600A048), ANIMMODE_ONCE,
+                         -3.0f);
+        this->animation = &D_0600A048;
+        this->animFunc = EnGe1_StopFidget;
+        this->stateFlags &= ~(GE1_STATE_IDLE_ANIM);
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_Ge1/z_en_ge1.c:340]:
```suggestion
        this->stateFlags &= ~GE1_STATE_IDLE_ANIM;
```

File: src/overlays/actors/ovl_En_Ge1/z_en_ge1.c [hunk 12]
```diff
@@ -46,95 +65,732 @@ static ColliderCylinderInit D_80A32770 = {
...
+void EnGe1_SetupOpenGate_GateOp(EnGe1* this, GlobalContext* globalCtx) {
+    this->stateFlags |= GE1_STATE_TALKING;
+
+    if ((func_8010BDBC(&globalCtx->msgCtx) == 5) && func_80106BC8(globalCtx)) {
+        this->actionFunc = EnGe1_OpenGate_GateOp;
+        Animation_Change(&this->skelAnime, &D_0600A048, 1.0f, 0.0f, Animation_GetLastFrame(&D_0600A048), ANIMMODE_ONCE,
+                         -3.0f);
+        this->animation = &D_0600A048;
+        this->animFunc = EnGe1_StopFidget;
+        this->stateFlags &= ~(GE1_STATE_IDLE_ANIM);
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_Ge1/z_en_ge1.c:435]:
```suggestion
        this->stateFlags &= ~GE1_STATE_IDLE_ANIM;
```

File: src/overlays/actors/ovl_En_Ge1/z_en_ge1.c [hunk 13]
```diff
@@ -46,95 +65,732 @@ static ColliderCylinderInit D_80A32770 = {
...
+        EnGe1_SetAnimationIdle(this);
+    }
+}
+
+void EnGe1_WaitTillItemGiven_Archery(EnGe1* this, GlobalContext* globalCtx) {
+    s32 getItemId;
+
+    if (Actor_HasParent(&this->actor, globalCtx)) {
+        this->actionFunc = EnGe1_SetupWait_Archery;
+        if ((this->stateFlags & GE1_STATE_GIVE_QUIVER)) {
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_Ge1/z_en_ge1.c:489]:
```suggestion
        if (this->stateFlags & GE1_STATE_GIVE_QUIVER) {
```

File: src/overlays/actors/ovl_En_Ge1/z_en_ge1.c [hunk 14]
```diff
@@ -46,95 +65,732 @@ static ColliderCylinderInit D_80A32770 = {
...
+        }
+        func_8002F434(&this->actor, globalCtx, getItemId, 10000.0f, 50.0f);
+    }
+}
+
+void EnGe1_BeginGiveItem_Archery(EnGe1* this, GlobalContext* globalCtx) {
+    s32 getItemId;
+
+    if (func_8002F334(&this->actor, globalCtx)) {
+        this->actor.flags = this->actor.flags & ~0x10000;
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_Ge1/z_en_ge1.c:516]:
```suggestion
        this->actor.flags &= ~0x10000;
```

File: src/overlays/actors/ovl_En_Ge1/z_en_ge1.c [hunk 15]
```diff
@@ -46,95 +65,732 @@ static ColliderCylinderInit D_80A32770 = {
...
+        gSaveContext.highScores[0] = gSaveContext.minigameScore;
+    }
+
+    if (gSaveContext.minigameScore < 1000) {
+        this->actor.textId = 0x6045;
+        this->actionFunc = EnGe1_TalkNoPrize_Archery;
+    } else if (!(gSaveContext.infTable[25] & 1)) {
+        this->actor.textId = 0x6046;
+        this->actionFunc = EnGe1_TalkWinPrize_Archery;
+        this->stateFlags &= ~(GE1_STATE_GIVE_QUIVER);
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_Ge1/z_en_ge1.c:631]:
```suggestion
        this->stateFlags &= ~GE1_STATE_GIVE_QUIVER;
```

File: src/overlays/actors/ovl_En_Ge1/z_en_ge1.c [hunk 16]
```diff
@@ -46,95 +65,732 @@ static ColliderCylinderInit D_80A32770 = {
...
+    Collider_UpdateCylinder(&this->actor, &this->collider);
+    CollisionCheck_SetOC(globalCtx, &globalCtx->colChkCtx, &this->collider.base);
+    Actor_MoveForward(&this->actor);
+    Actor_UpdateBgCheckInfo(globalCtx, &this->actor, 40.0f, 25.0f, 40.0f, 5);
+    this->animFunc(this);
+    this->actionFunc(this, globalCtx);
+
+    if (this->stateFlags & GE1_STATE_TALKING) {
+        EnGe1_TurnToFacePlayer(this, globalCtx);
+        this->stateFlags &= ~(GE1_STATE_TALKING);
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_Ge1/z_en_ge1.c:719]:
```suggestion
        this->stateFlags &= ~GE1_STATE_TALKING;
```

File: src/overlays/actors/ovl_En_Ge1/z_en_ge1.c [hunk 17]
```diff
@@ -46,95 +65,732 @@ static ColliderCylinderInit D_80A32770 = {
...
+    s32 pad;
+    EnGe1* this = THIS;
+
+    if (limbIndex == GE1_LIMB_HEAD) {
+        rot->x += this->headRot.y;
+        rot->z += this->headRot.x;
+    }
+
+    if (this->stateFlags & GE1_STATE_STOP_FIDGET) {
+        this->stateFlags &= ~(GE1_STATE_STOP_FIDGET);
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_Ge1/z_en_ge1.c:762]:
```suggestion
        this->stateFlags &= ~GE1_STATE_STOP_FIDGET;
```

**Comment by fig02** [comment]
changes made

**Review by Roman971 — APPROVED** [review]

**Pull request merged**